staging/src/k8s.io/client-go/rest/request_watchlist_test.go

```diff
@@ -0,0 +1,337 @@
+			}(scenario.watchEvents)
+
+			res := target.handleWatchList(ctx, fakeWatcher)
+			if res.err != nil {
+				t.Fatal(res.err)
+			}
+
+			result := &v1.PodList{}
+			if err := res.Into(result); err != nil {
+				t.Fatal(err)
+			}
+			if !apiequality.Semantic.DeepEqual(scenario.expectedResult, result) {
+				t.Errorf("diff: %v", cmp.Diff(scenario.expectedResult, result))
+			}
+			if !fakeWatcher.IsStopped() {
+				t.Fatalf("the watcher wasn't stopped")
+			}
+		})
+	}
+}
+
+func TestWatchListFailure(t *testing.T) {
+	scenarios := []struct {
+		name        string
+		ctx         context.Context
+		watcher     *watch.FakeWatcher
+		watchEvents []watch.Event
+
+		expectedError error
+	}{
+		{
+			name: "request stop",
+			ctx: func() context.Context {
+				ctx, ctxCancel := context.WithCancel(context.TODO())
+				ctxCancel()
+				return ctx
+			}(),
+			expectedError: fmt.Errorf("context canceled"),
+		},
+		{
+			name: "stop watcher",
+			watcher: func() *watch.FakeWatcher {
+				w := watch.NewFake()
+				w.Stop()
+				return w
+			}(),
+			expectedError: fmt.Errorf("unexpected watch close"),
+		},
+		{
+			name:          "stop on watch.Error",
+			watchEvents:   []watch.Event{{Type: watch.Error, Object: &apierrors.NewInternalError(fmt.Errorf("dummy errror")).ErrStatus}},
+			expectedError: fmt.Errorf("Internal error occurred: dummy errror"),
+		},
+		{
+			name:          "incorrect watch type (Deleted)",
+			watchEvents:   []watch.Event{{Type: watch.Deleted, Object: makePod(1)}},
+			expectedError: fmt.Errorf("unexpected watch event .*, expected to only receive watch.Added and watch.Bookmark events"),
+		},
+		{
+			name:          "incorrect watch type (Modified)",
+			watchEvents:   []watch.Event{{Type: watch.Modified, Object: makePod(1)}},
+			expectedError: fmt.Errorf("unexpected watch event .*, expected to only receive watch.Added and watch.Bookmark events"),
+		},
+		{
+			name:          "unordered input returns an error",
+			watchEvents:   []watch.Event{{Type: watch.Added, Object: makePod(3)}, {Type: watch.Added, Object: makePod(1)}},
+			expectedError: fmt.Errorf("cannot add the obj .* with the key = ns/pod-1, as it violates the ordering guarantees provided by the watchlist feature in beta phase, lastInsertedKey was = ns/pod-3"),
+		},
+	}
+
+	for _, scenario := range scenarios {
+		t.Run(scenario.name, func(t *testing.T) {
+			target := &Request{}
+			if scenario.watcher == nil {
+				scenario.watcher = watch.NewFake()
+			}
+			if scenario.ctx == nil {
+				scenario.ctx = context.TODO()
+			}
+			go func(w *watch.FakeWatcher, watchEvents []watch.Event) {
+				for _, event := range watchEvents {
+					w.Action(event.Type, event.Object)
+				}
+			}(scenario.watcher, scenario.watchEvents)
+
+			res := target.handleWatchList(scenario.ctx, scenario.watcher)
+			resErr := res.Into(nil)
+			if resErr == nil {
+				t.Fatal("expected to get an error")
+			}
+			matched, err := regexp.MatchString(scenario.expectedError.Error(), resErr.Error())
```

Comment:
can't we just compare strings? Why do we need regexp here?

Reply:
the regexp allows to define the expected error as:

`cannot add the obj .* with the key = ns/pod-1, as it violates the ordering guarantees provided by the watchlist feature in beta phase, lastInsertedKey was = ns/pod-3`

instead of:

`cannot add the obj (&v1.Pod{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"pod-1", GenerateName:"", Namespace:"ns", SelfLink:"", UID:"", ResourceVersion:"1", Generation:0, CreationTimestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), DeletionTimestamp:<nil>, DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string{}, OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ManagedFields:[]v1.ManagedFieldsEntry(nil)}, Spec:v1.PodSpec{Volumes:[]v1.Volume(nil), InitContainers:[]v1.Container(nil), Containers:[]v1.Container(nil), EphemeralContainers:[]v1.EphemeralContainer(nil), RestartPolicy:"", TerminationGracePeriodSeconds:(*int64)(nil), ActiveDeadlineSeconds:(*int64)(nil), DNSPolicy:"", NodeSelector:map[string]string(nil), ServiceAccountName:"", DeprecatedServiceAccount:"", AutomountServiceAccountToken:(*bool)(nil), NodeName:"", HostNetwork:false, HostPID:false, HostIPC:false, ShareProcessNamespace:(*bool)(nil), SecurityContext:(*v1.PodSecurityContext)(nil), ImagePullSecrets:[]v1.LocalObjectReference(nil), Hostname:"", Subdomain:"", Affinity:(*v1.Affinity)(nil), SchedulerName:"", Tolerations:[]v1.Toleration(nil), HostAliases:[]v1.HostAlias(nil), PriorityClassName:"", Priority:(*int32)(nil), DNSConfig:(*v1.PodDNSConfig)(nil), ReadinessGates:[]v1.PodReadinessGate(nil), RuntimeClassName:(*string)(nil), EnableServiceLinks:(*bool)(nil), PreemptionPolicy:(*v1.PreemptionPolicy)(nil), Overhead:v1.ResourceList(nil), TopologySpreadConstraints:[]v1.TopologySpreadConstraint(nil), SetHostnameAsFQDN:(*bool)(nil), OS:(*v1.PodOS)(nil), HostUsers:(*bool)(nil), SchedulingGates:[]v1.PodSchedulingGate(nil), ResourceClaims:[]v1.PodResourceClaim(nil)}, Status:v1.PodStatus{Phase:"", Conditions:[]v1.PodCondition(nil), Message:"", Reason:"", NominatedNodeName:"", HostIP:"", HostIPs:[]v1.HostIP(nil), PodIP:"", PodIPs:[]v1.PodIP(nil), StartTime:<nil>, InitContainerStatuses:[]v1.ContainerStatus(nil), ContainerStatuses:[]v1.ContainerStatus(nil), QOSClass:"", EphemeralContainerStatuses:[]v1.ContainerStatus(nil), Resize:"", ResourceClaimStatuses:[]v1.PodResourceClaimStatus(nil)}}) with the key = ns/pod-1, as it violates the ordering guarantees provided by the watchlist feature in beta phase, lastInsertedKey was = ns/pod-3`

Reply:
in the reflector we print the entire object for troubleshooting purposes. alternatively we could change the error msg to not contain the whole object.

Reply:
no that's ok. Let's leave as is